add: Crawlee blog (#4)
Arindam200/LinkedIn_Scraping: article.md

# How to Build a LinkedIn Job Posting Scraper using Crawlee for Python and Streamlit

## Introduction

In this article, we are going to be building a web application to scrape LinkedIn’s website for job postings using Crawlee and Streamlit.

We will create a LinkedIn job scraper utilising Crawlee for Python to extract the company name, job title, time of posting and link to the job posting, from dynamically received user input through the web application.

By the end of this tutorial, you’ll have had a fully functional web application that you can use to scrape job postings from the LinkedIn website.

![alt_text](images/image1.png "image_tooltip")

> 
> Let’s get started
> 
> 

Reviewer: ```suggestion
```

This is redundant, we say "Let's start" on the next line"

## Prerequisites

Let's start by creating a new Crawlee for Python project with this command:

```
pipx run crawlee create linkedin-scraper
```

Select `PlaywrightCrawler` in the terminal when Crawlee asks for it.

After installation, Crawlee for Python will create a boilerplate code for you. Redirect into the project folder and run this command to install dependencies.

```
poetry install
```

## Building the LinkedIn Scraper using Crawlee for Python

> In this section, we are going to be building the scraper using the Crawlee Python package. To learn more about Crawlee, check out their [documentation](https://crawlee.dev/python/docs/quick-start).

Reviewer: ```suggestion
In this section, we are going to be building the scraper using the Crawlee for Python package. To learn more about Crawlee, check out their [documentation](https://crawlee.dev/python/docs/quick-start).
```

### Inspecting the LinkedIn job Search Page

Open LinkedIn on your web browser and sign out from the website (if you already have an account logged in). You should see an interface like this.

![alt_text](images/image2.png "image_tooltip")

> Navigate to the jobs section, search for a job and location of your choice; copy the Uniform Resource Locator (URL).

Reviewer: ```suggestion
Navigate to the jobs section, search for a job and location of your choice; copy the URL.
```

We're going to focus on the search parameters, which is the part that goes after '?'. The keyword and location parameters are the most important ones for us. 

The job title the user supplies will serve as input to the keyword parameter while the location the user supplies will go into the location parameter. Lastly, the geoId parameter will be removed while we keep the other parameters constant.

We are going to be making changes to our `main.py` file. Copy and paste the code below in your `main.py` file.

```
from crawlee.playwright_crawler import PlaywrightCrawler
from .routes import router                                     
import urllib.parse

async def main(title: str, location: str, data_name: str) -> None:
    base_url = "https://www.linkedin.com/jobs/search"

    # URL encode the parameters
    params = {
        "keywords": title,
        "location": location,
        "trk": "public_jobs_jobs-search-bar_search-submit",
        "position": "1",
        "pageNum": "0"
    }
    encoded_params = urllib.parse.urlencode(params)

>     encoded_url = f"{base_url}?{encoded_params}"

Reviewer: I'm not a Pythonista, but doesn't urllib also have a function to resolve base URLs? It feels wrong to use a specialized URL manipulation library for params, and then manually stitch the strings together with the base URL.

    # Initialize the crawler
    crawler = PlaywrightCrawler(
        request_handler=router,

>     )

Reviewer: This is kinda confusing, because we haven't created or shown the router file yet, or explained what it is. The readers might think this is a runnable example and they'll be left with errors without any explanation.

Please start with some intermediary runnable state, (aka console.log in request handler), guide the user to making sure the crawler logs correctly, and then tell them that we will replace it with routing.

Author: During installation of the crawlee package, a boilerplate code has been created that contains the router.py file

We are simply editing those files.

By the time the user is done with the setup section, he will have the boiler plate code and he will see the routes.py file

Reviewer: Ah yes, forgot we're starting from the template. 👍 

Still, would be nice to show the user what the correct behavior is, so that they feel comfortable they've set up everything correctly.

Usually I do something like:

> When you run the code, you will see the following output:

> [screenshot of output from terminal]

Author: Okay, I've updated that as well.

Please check now

    # Run the crawler with the initial list of URLs
    await crawler.run([encoded_url])

    # Save the data in a CSV file
    output_file = f"{data_name}.csv"
    await crawler.export_data(output_file)
```

 

> Now that we have encoded the URL, the next step for us is to adjust the generated router to handle linkedin job postings. 

Reviewer: ```suggestion
Now that we have encoded the URL, the next step for us is to adjust the generated router to handle LinkedIn job postings. 
```

### Routing your Crawler

We will be making use of two handlers for your application:

* **Default handler** 

The `default_handler` handles the start URL

* **Job listing**

The `job_listing` handler extracts the individual job details.

> The Playwright crawler is going to crawl through the job posting page and extract the links to all job postings on the page.

Reviewer: ```suggestion
PlaywrightCrawler is going to crawl through the job posting page and extract the links to all job postings on the page.
```

    #add all the links to the job listing route
    await context.add_requests(
            [Request.from_url(rec, label='job_listing') for rec in hrefs]
        )
```

Now that we have the job listings, the next step is to scrape their details.

We'II extract each job’s title, company's name, time of posting and the link to the job post. Open your dev tools to extract each element Xpath.

![alt_text](images/image5.png "image_tooltip")

After scraping each of the listings, we'll remove special characters from the text to make it clean and push the data to local storage using `context.push_data` function.

```
@router.handler('job_listing')
async def listing_handler(context: PlaywrightCrawlingContext) -> None:
    """Handler for job listings."""

    await context.page.wait_for_load_state('load')

    job_title = await context.page.locator('//*[@id="main-content"]/section[1]/div/section[2]/div/div[1]/div/h1').text_content()

    company_name  = await context.page.locator('//*[@id="main-content"]/section[1]/div/section[2]/div/div[1]/div/h4/div[1]/span[1]/a').text_content()

>     time_of_posting= await context.page.locator('//*[@id="main-content"]/section[1]/div/section[2]/div/div[1]/div/h4/div[2]/span').text_content()

Reviewer: Those are super brittle selectors. A single change in any of those paths will break them. We need to find more stable ones. There are multiple different ways to find elements using Playwright and XPath is usually the worst of them.

    await context.push_data(
        {
            'title': re.sub(r'[\s\n]+', '', job_title),
            'Company name': re.sub(r'[\s\n]+', '', company_name),

>             'Time of posting': re.sub(r'[\s\n]+', '', time_of_posting),

Reviewer: A comment why we need to do those regex replacements would be helpful

            'url': context.request.loaded_url,
        }
    )
```

## Creating your Application

For this project, we will be using Streamlit for the web application. Before we proceed, we are going to create a new file named `app.py` in your project directory. In addition, ensure you have  [Streamlit](https://docs.streamlit.io/get-started/installation) installed in your global Python environment before proceeding with this section.

```
import streamlit as st
import subprocess

# Streamlit form for inputs 
st.title("LinkedIn Job Scraper")

with st.form("scraper_form"):
    title = st.text_input("Job Title", value="backend developer")
    location = st.text_input("Job Location", value="newyork")

>     data_name = st.text_input("Output File Name", value="backend_job")

Reviewer: ```suggestion
    data_name = st.text_input("Output File Name", value="backend_jobs")
```

it will be more than one, right?

# How to Build a LinkedIn Job Posting Scraper using Crawlee for Python and Streamlit

## Introduction

In this article, we are going to be building a web application to scrape LinkedIn’s website for job postings using Crawlee and Streamlit.

We will create a LinkedIn job scraper utilising Crawlee for Python to extract the company name, job title, time of posting and link to the job posting, from dynamically received user input through the web application.

By the end of this tutorial, you’ll have had a fully functional web application that you can use to scrape job postings from the LinkedIn website.

![alt_text](images/image1.png "image_tooltip")

Let’s get started

## Prerequisites

Let's start by creating a new Crawlee for Python project with this command:

```
pipx run crawlee create linkedin-scraper
```

Select `PlaywrightCrawler` in the terminal when Crawlee asks for it.

After installation, Crawlee for Python will create a boilerplate code for you. Redirect into the project folder and run this command to install dependencies.

```

> Poetry install

Reviewer 2: ```suggestion
poetry install
```

    output_file = f"{data_name}.csv"
    await crawler.export_data(output_file)
```

 

Now that we have encoded the URL, the next step for us is to adjust the generated router to handle linkedin job postings. 

### Routing your Crawler

We will be making use of two handlers for your application:

* **Default handler** 

The `default_handler` handles the start URL

* **Job listing**

The `job_listing` handler extracts the individual job details.

The Playwright crawler is going to crawl through the job posting page and extract the links to all job postings on the page.

![alt_text](images/image4.png "image_tooltip")

When you examine the job postings, you will discover that the job posting links are inside an ordered list with a class named `jobs-search__results-list`. We will then extract the links using the Playwright locator object and add them to the `job_listing` route for processing.

```
from contextlib import suppress, asynccontextmanager
from crawlee.router import Router
from crawlee import Request
from crawlee.playwright_crawler import PlaywrightCrawlingContext
from playwright.async_api import TimeoutError as PlaywrightTimeoutError

> import re

Reviewer 2: There are some unused imports here.

# How to Build a LinkedIn Job Posting Scraper using Crawlee for Python and Streamlit

## Introduction

In this article, we are going to be building a web application to scrape LinkedIn’s website for job postings using Crawlee and Streamlit.

We will create a LinkedIn job scraper utilising Crawlee for Python to extract the company name, job title, time of posting and link to the job posting, from dynamically received user input through the web application.

By the end of this tutorial, you’ll have had a fully functional web application that you can use to scrape job postings from the LinkedIn website.

![alt_text](images/image1.png "image_tooltip")

Let’s get started

## Prerequisites

> Let's start by installing Crawlee for Python with this command:

Reviewer 2: ```suggestion
Let's start by creating a new Crawlee for Python project with this command:
```

```
pipx run crawlee create linkedin-scraper
```

Select `PlaywrightCrawler` in the terminal when Crawlee asks for it.

After installation, Crawlee for Python will create a boilerplate code for you. Redirect into the project folder and run this command to install dependencies.

```
Poetry install
```

## Building the LinkedIn Scraper using Crawlee for Python

In this section, we are going to be building the scraper using the Crawlee Python package. To learn more about Crawlee, check out their [documentation](https://crawlee.dev/python/docs/quick-start).

				

### Inspecting the LinkedIn job Search Page

Open LinkedIn on your web browser and sign out from the website (if you already have an account logged in). You should see an interface like this.

![alt_text](images/image2.png "image_tooltip")

> Navigate to the jobs section, search for a job and location of your choice; copy the Uniform Resource Locator (URL).

Reviewer 2: I believe it's OK to just say "URL" without explanation

![alt_text](images/image3.png "image_tooltip")

You should have something like this:

`https://www.linkedin.com/jobs/search?keywords=Backend%20Developer&location=Canada&geoId=101174742&trk=public_jobs_jobs-search-bar_search-submit&position=1&pageNum=0` 

The URL can be divided into two:

* The base URL
* The parameters

> The base URL is usually before the question mark while the parameters come after the question mark. Furthermore, you can change the inputs to the parameters to get different results on your webpage.

Reviewer 2: I don't think this is correct terminology. I suggest rephrasing this so that it just says something like "We're going to focus on the search parameters, which is the part that goes after '?'. The keyword and location parameters are the most important ones for us".

I'd skip the explanation of base url/parameters completely.

```
from crawlee.playwright_crawler import PlaywrightCrawler
from .routes import router                                     
import urllib.parse

async def main(title: str, location: str, data_name: str) -> None:
    base_url = "https://www.linkedin.com/jobs/search"

    # URL encode the parameters
    params = {
        "keywords": title,
        "location": location,
        "trk": "public_jobs_jobs-search-bar_search-submit",
        "position": "1",
        "pageNum": "0"
    }
    encoded_params = urllib.parse.urlencode(params)
    encoded_url = f"{base_url}?{encoded_params}"

    # Initialize the crawler
    crawler = PlaywrightCrawler(
        request_handler=router,
    )

    # Run the crawler with the initial list of URLs
    await crawler.run([encoded_url])

    # Save the data in a CSV file
    output_file = f"{data_name}.csv"
    await crawler.export_data(output_file)
```

 

The main function takes in the title, location and data_name parameters. Furthermore, we used the [Urllib](https://docs.python.org/3/library/urllib.html) library to encode the base URL and its parameters into a format that can be used to make requests to the web. 

> Now that we have encoded the URL, the next step for us is to set up the playwright crawler to take in your custom-defined routers. 

Reviewer 2: That's already done in the generated project - in fact, the next step is adjusting the generated router to handle linkedin job postings.

### Routing your Crawler

> We will be making use of two routers for your application. These routes include:

Reviewer 2: ```suggestion
We will be making use of two handlers for your application:
```

* **Default handler** 

> The `default_handler` route is the fallback route that handles the start URL

Reviewer 2: ```suggestion
The `default_handler` handles the start URL
```

* **Job listing**

> The `job_listing` route extracts the individual job details.

Reviewer 2: ```suggestion
The `job_listing` handler extracts the individual job details.
```

![alt_text](images/image4.png "image_tooltip")

When you examine the job postings, you will discover that the job posting links are inside an ordered list with a class named `jobs-search__results-list`. We will then extract the links using the Playwright locator object and add them to the `job_listing` route for processing.

```
from contextlib import suppress, asynccontextmanager
from crawlee.router import Router
from crawlee import Request
from crawlee.playwright_crawler import PlaywrightCrawlingContext
from playwright.async_api import TimeoutError as PlaywrightTimeoutError
import re

router = Router[PlaywrightCrawlingContext]()

@router.default_handler
async def default_handler(context: PlaywrightCrawlingContext) -> None:
    """Default request handler."""

    #select all the links for the job posting on the page
    hrefs = await context.page.locator('ul.jobs-search__results-list a').evaluate_all("links => links.map(link => link.href)")

    #add all the links to the job listing route
    await context.add_requests(
            [Request.from_url(rec, label='job_listing') for rec in hrefs]
        )
```

> Now that we have the job postings, the next step is to scrape the details on them.

Reviewer 2: ```suggestion
Now that we have the job listings, the next step is to scrape their details.
```

We'II extract each job’s title, company's name, time of posting and the link to the job post. Open your dev tools to extract each element Xpath.

![alt_text](images/image5.png "image_tooltip")

> After scraping each of the parameters, we'll remove special characters from the text to make it clean and push the data to local storage using `context.push_data` function.

Reviewer 2: ```suggestion
After scraping each of the listings, we'll remove special characters from the text to make it clean and push the data to local storage using `context.push_data` function.
```

![alt_text](images/image1.png "image_tooltip")

> Let's start

Reviewer: The point was to delete this. Sorry if that wasn't clear.

```suggestion
```

poetry install
```

> We are going to begin editing the files provided to us by crawlee in order for us to build our scraper

Reviewer: ```suggestion
We are going to begin by editing the files provided to us by Crawlee in order for us to build our scraper.
```

        "position": "1",
        "pageNum": "0"
    }

>     encoded_params = urlencode(params)

Reviewer: AFAIK this is not executable, because urlencode is not imported.

    
    # Encode parameters into a query string
    query_string = '?' + encoded_params

    # Combine base URL with the encoded query string

>     encoded_url = urljoin(base_url, "") + query_string

Reviewer: IMO this is not better than the previous solution. If there's no nice way to do what I wanted in Python, you could've just said so ;) Let's use the original version.

The `job_listing` handler extracts the individual job details.

> Playwright crawler is going to crawl through the job posting page and extract the links to all job postings on the page.

Reviewer: ```suggestion
PlaywrightCrawler is going to crawl through the job posting page and extract the links to all job postings on the page.
```

Reviewer: I have a feeling I already fixed that once...

    await context.page.wait_for_load_state('load')

    job_title = await context.page.locator('div.top-card-layout__entity-info h1.top-card-layout__title').text_content()
   
    company_name  = await context.page.locator('span.topcard__flavor a').text_content()   

    time_of_posting= await context.page.locator('div.topcard__flavor-row span.posted-time-ago__text').text_content()

    await context.push_data(
        {

>             # we are maing use of regex to remove special characters for the extracted texts

Reviewer: Even if I ignore the fact that the sentence is terribly misspelled...

The question was, why do we need to remove special characters from the texts?  The fact that the regex removes special characters is clear. The point of comments is to explain what the reader cannot see from the code.
